# Chapter 1

### 1.1 First Program: "Hello World"

A classic first program, "Hello World," simply prints a message to the console. This demonstrates how to use the `print()` function.

In [ ]:
print("Hello World")

Hello World


## Working with Libraries: Jokes!

Python's power comes from its vast collection of libraries. We'll explore how to use external libraries to add functionality to our code.

The `pyjokes` library provides a simple way to get programmer jokes. Since it's not a built-in library, we first need to install it using `!pip install`. The `!` at the beginning tells Colab to run the command as a shell command.

In [ ]:
!pip install pyjokes

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.6/47.6 kB 3.4 MB/s eta 0:00:00


Once installed, we can import `pyjokes` and use its functions, like `get_joke()`, to fetch a random joke.

In [ ]:
import pyjokes

# This prints a random joke
joke = pyjokes.get_joke()
print(joke)

I have a joke on programming but it only works on my computer.


# chapter 2 `Variables and data types`

### 1. variables
### 1.1 Declaring and Assigning Variables

You declare a variable by giving it a name and assigning it a value using the `=` operator.

In [ ]:
a = 1

b = 2

c = 7

name = "rohan"

print(a + b)

3


### 2. data types
### 2.1 Common Data Types

Here are examples of some fundamental Python data types:
- **Integer (`int`)**: Whole numbers (e.g., `1`, `100`, `-5`).
- **Floating-point number (`float`)**: Numbers with decimal points (e.g., `3.14`, `0.001`, `-2.5`).
- **String (`str`)**: Sequences of characters, enclosed in single or double quotes (e.g., `"Hello"`, `'Python'`).
- **Boolean (`bool`)**: Represents truth values, either `True` or `False`.
- **NoneType (`None`)**: Represents the absence of a value.

In [ ]:
a = 1 # a is an integer

b = 5.22 # b is a floating point number

c = "sahil" # c is a string

d = False # d is a boolean variable

e = None # e is a none type variable

### 3. rules for variables
### 3.1 Valid and Invalid Variable Names

*   Must start with a letter (a-z, A-Z) or an underscore (`_`).
*   Cannot start with a number.
*   Can only contain alpha-numeric characters (a-z, A-Z, 0-9) and underscores (`_`).
*   Variable names are case-sensitive (`age` is different from `Age`).
*   Cannot be a Python keyword (e.g., `if`, `for`, `while`).

 @sameer = 56 # Invalid due to @ symbol

 s@meer # Invalid due to @ symbol

In [ ]:

a = 23

aaa = 435

harry = 34

sameer = 45

_samerr = 34


### 4. Operators
### 4.1 Arithmetic Operators

These operators perform mathematical calculations like addition, subtraction, multiplication, and division.

### 4.2 Assignment Operators

These operators are used to assign values to variables. They combine an arithmetic operation with assignment.

### 5.3 Comparison (Relational) Operators

These operators compare two values and return a Boolean result (`True` or `False`).

### 5.4 Logical Operators

These operators combine conditional statements (`True`/`False` expressions) and are used to evaluate logical relationships.
- `and`: Returns `True` if both statements are true.
- `or`: Returns `True` if at least one statement is true.
- `not`: Reverses the result; returns `False` if the result is true.

In [ ]:
# Arithmetic Operators
a = 7
b = 4
c = a + b
print(c)

11


Let's explore all the common arithmetic operators with more examples:

In [ ]:
# Arithmetic Operators in action
num1 = 10
num2 = 3

print(f"Addition: {num1} + {num2} = {num1 + num2}")
print(f"Subtraction: {num1} - {num2} = {num1 - num2}")
print(f"Multiplication: {num1} * {num2} = {num1 * num2}")
print(f"Division: {num1} / {num2} = {num1 / num2}")
print(f"Modulus (Remainder): {num1} % {num2} = {num1 % num2}")
print(f"Exponentiation: {num1} ** {num2} = {num1 ** num2}")
print(f"Floor Division (Integer Division): {num1} // {num2} = {num1 // num2}")

Addition: 10 + 3 = 13
Subtraction: 10 - 3 = 7
Multiplication: 10 * 3 = 30
Division: 10 / 3 = 3.3333333333333335
Modulus (Remainder): 10 % 3 = 1
Exponentiation: 10 ** 3 = 1000
Floor Division (Integer Division): 10 // 3 = 3


In [ ]:
# Assignment Operators
a = 4-2 # Assign 4-2 in a
print(a)
b = 6
# b += 3 # Increment the value of b by 3 and then assign it to b
b -= 3 # Decrement the value of b by 3 and then assign it to b
print(b)

2
3


Here are more examples of assignment operators that combine an arithmetic operation with assignment:

In [ ]:
# Assignment Operators in action
x = 10
print(f"Initial x: {x}")

x += 5 # Equivalent to x = x + 5
print(f"x after x += 5: {x}")

x -= 3 # Equivalent to x = x - 3
print(f"x after x -= 3: {x}")

x *= 2 # Equivalent to x = x * 2
print(f"x after x *= 2: {x}")

x /= 4 # Equivalent to x = x / 4
print(f"x after x /= 4: {x}")

x %= 3 # Equivalent to x = x % 3
print(f"x after x %= 3: {x}")

x = 2 # Reset x for exponentiation example
x **= 3 # Equivalent to x = x ** 3
print(f"x after x **= 3: {x}")

x = 10 # Reset x for floor division example
x //= 3 # Equivalent to x = x // 3
print(f"x after x //= 3: {x}")

Initial x: 10
x after x += 5: 15
x after x -= 3: 12
x after x *= 2: 24
x after x /= 4: 6.0
x after x %= 3: 0.0
x after x **= 3: 8
x after x //= 3: 3


In [ ]:
# Comparison Operators

d = 5==5

print(d)

True


Let's look at a full range of comparison operators and their results:

In [ ]:
# Comparison Operators in action
p = 15
q = 20
r = 15

print(f"{p} == {q}: {p == q}") # Equal to
print(f"{p} == {r}: {p == r}")
print(f"{p} != {q}: {p != q}") # Not equal to
print(f"{p} > {q}: {p > q}")   # Greater than
print(f"{p} < {q}: {p < q}")   # Less than
print(f"{p} >= {r}: {p >= r}") # Greater than or equal to
print(f"{q} <= {p}: {q <= p}") # Less than or equal to

15 == 20: False
15 == 15: True
15 != 20: True
15 > 20: False
15 < 20: True
15 >= 15: True
20 <= 15: False


In [ ]:
# Logical Operators

e = True or False

# Truth table of 'or'
print("True or False is ", True or False)
print("True or True is ", True or True)
print("False or True is ", False or True)
print("False or False is ", False or False)

# Truth table of 'and'
print("True and False is ", True and False)
print("True and True is ", True and True)
print("False and True is ", False and True)
print("False and False is ", False and False)

print(not(True))

True or False is  True
True or True is  True
False or True is  True
False or False is  False
True and False is  False
True and True is  True
False and True is  False
False and False is  False
False


### 5. Type of data

In [ ]:
a = "31.2"
b = float(a) # a but the type should be float
t = type(b)

print(t)

<class 'float'>


### 5. input function

In [ ]:
a = int(input("Enter number 1: "))

b = int(input("Enter number 2: "))

print("Number a is: ", a)
print("Number b is: ", b)
print("Sum is ", a + b)

Enter number 1: 5
Enter number 2: 5
Number a is:  5
Number b is:  5
Sum is  10


##Qusetions

1. Write a python program to add two numbers.

In [ ]:
a = 1

b = 5

print(a + b)

6


2. Write a python program to find remainder when a number is divided by z.

In [ ]:
a = 37

b = 5

print("Remainder when a is divided by b is", a % b)

Remainder when a is divided by b is 2


3. Check the type of variable assigned using input () function.

In [ ]:
a = input("Enter the value of a: ")
print(type(a))

Enter the value of a: hello
<class 'str'>


4. Use comparison operator to find out whether `a` given variable a is greater than
`b` or not. Take a = 34 and b = 80


In [ ]:
a = int(input("Enter number 1: "))
b = int(input("Enter number 2: "))

print("a is greater than b is", a>b)

Enter number 1: 7
Enter number 2: 4
a is greater than b is True


5. Write a python program to find an average of two numbers entered by the user.

In [ ]:
a = int(input("Enter number 1: "))
b = int(input("Enter number 2: "))


print("The average of these two number is", (a+b)/2)

Enter number 1: 8
Enter number 2: 7
The average of these two number is 7.5


6. Write a python program to calculate the square of a number entered by the user.

In [ ]:
a = int(input("Enter your number: "))


print("The square of the number is", a**2)
print("The square of the number is", a*a)

Enter your number: 7
The square of the number is 49
The square of the number is 49


#Chapter 3 `strings`

String is a data type in python.

String is a sequence of characters enclosed in quotes.

We can primarily write a string in these three ways.

In [ ]:
a ='hello' # Single quoted string
b = "hello" # Double quoted string
c = '''hello''' # Triple quoted string

string slicing

In [ ]:
name = "Rohan"

nameshort = name[0:3] # start from index 0 all the way till 3 (excluding 3)
print(nameshort)
character1 = name[1]
print(character1)

Roh
o


negative slicing

In [ ]:
name = "Rohan"

print(name[0:3])

print(name[-4: -1])
print(name[1:4])

print(name[:4]) # is same as print(name[0:4])
print(name[1:]) # is same as print(name[1:5])
print(name[1:5])

Roh
oha
oha
Roha
ohan
ohan


SLICING WITH SKIP VALUE
We can provide a skip value as a part of our slice like this:

In [ ]:
word = "amazing"
word[1: 6: 2] # "mzn"

'mzn'

Other advanced slicing techniques:

In [ ]:
Word = "amazing"
Word[:7] # word [0:7] – 'amazing'
Word[0:] # word [0:7] – 'amazing'

'amazing'

###functions for string

In [ ]:
name = "Avengers"

1. len () function – This function returns the length of the strings.

In [ ]:
print(len(name))

8


2. String.endswith("ers") – This function_ tells whether the variable string ends with
the string "ers" or not. If string is "Avengers", it returns true for "ers" since Avengers ends with ers.

In [ ]:
print(name.endswith("ers"))

True


In [ ]:
print(name.startswith("Av"))

True


3. string.count("c") – counts the total number of occurrences of any character.


In [ ]:
str = "google"
count = str.count("o")
print(count) # Output: 2

2


4. the first character of a given string.

In [ ]:
str = "google"
print(str.capitalize())

Google


5. string.find(word) – This function finds a word and returns the index of first
occurrence of that word in the string.

In [ ]:
str = "google"
index = str.find("og")
print(index) # Output: 2

2


6. string.replace (old word, new word ) – This function replace the old word with
new word in the entire string.

In [ ]:
my_str = "Hire"
replaced_string = my_str.replace("H", "F")
print(replaced_string)

Fire


### 7. String Concatenation

Concatenation is the process of joining two or more strings together. You can use the `+` operator to combine strings.

In [ ]:
first_name = "John"
last_name = "Doe"
full_name = first_name + " " + last_name
print(full_name)

greeting = "Hello" + ", " + "World!"
print(greeting)

John Doe
Hello, World!


### 8. F-strings (Formatted String Literals)

F-strings provide a concise and convenient way to embed Python expressions inside string literals for formatting. They are prefixed with `f` or `F`.

In [ ]:
name = "Alice"
age = 30

# Using f-string for formatting
message = f"My name is {name} and I am {age} years old."
print(message)

# You can also perform operations inside f-strings
print(f"The sum of 5 and 3 is {5 + 3}")

My name is Alice and I am 30 years old.
The sum of 5 and 3 is 8


### 9. String Immutability

In Python, strings are **immutable**. This means that once a string is created, its content cannot be changed. Any operation that appears to modify a string, like replacement or concatenation, actually creates a new string.

In [ ]:
my_string = "Python"
print(f"Original string: {my_string}")

# Trying to change a character will result in an error
# my_string[0] = 'J' # This would raise a TypeError

# Operations like replace create a new string
new_string = my_string.replace('P', 'J')
print(f"New string after replace: {new_string}")
print(f"Original string is still: {my_string}") # Original string remains unchanged

Original string: Python
New string after replace: Jython
Original string is still: Python


### 10. More Essential String Methods

Here are some other commonly used string methods that are very useful, especially in data preprocessing for Machine Learning:

In [ ]:
# .strip() - Removes leading and trailing whitespace
text = "  Hello, World!  "
stripped_text = text.strip()
print(f"strip(): '{stripped_text}'")

# .lower() - Converts all characters to lowercase
message = "HELLO Python"
lower_message = message.lower()
print(f"lower(): '{lower_message}'")

# .upper() - Converts all characters to uppercase
statement = "this is a test"
upper_statement = statement.upper()
print(f"upper(): '{upper_statement}'")

# .split() - Splits the string into a list of substrings based on a delimiter
sentence = "Python is a versatile language"
words = sentence.split()
print(f"split() by default (whitespace): {words}")

data = "apple,banana,cherry"
fruits = data.split(',')
print(f"split(',') with comma delimiter: {fruits}")

# .join() - Joins elements of an iterable (like a list) with the string as a separator
word_list = ['Join', 'these', 'words']
joined_string = ' '.join(word_list)
print(f"join() with space separator: '{joined_string}'")

dash_separated = '-'.join(['file', 'name', 'txt'])
print(f"join() with hyphen separator: '{dash_separated}'")

strip(): 'Hello, World!'
lower(): 'hello python'
upper(): 'THIS IS A TEST'
split() by default (whitespace): ['Python', 'is', 'a', 'versatile', 'language']
split(',') with comma delimiter: ['apple', 'banana', 'cherry']
join() with space separator: 'Join these words'
join() with hyphen separator: 'file-name-txt'


ESCAPE SEQUENCE CHARACTERS

Sequence of characters after backslash `"\" → Escape Sequence characters`

Escape Sequence characters comprise of more than one character but represent one character when used within the strings. Common escape sequences include:
- `\n`: Newline
- `\t`: Tab
- `\\`: Backslash
- `\'`: Single quote
- `\"`: Double quote

In [ ]:
a = 'nice to meet you \n my name is \'Yash\''

print(a)

nice to meet you 
 my name is 'Yash'


In [ ]:
path = "C:\\Users\\Yash\\Documents"
print(f"Path with backslashes: {path}")

Path with backslashes: C:\Users\Yash\Documents


In [ ]:
a = 'nice to meet you \n\tmy name is \'Yash\''
print(a)

nice to meet you 
	my name is 'Yash'


##Questions

In [ ]:
name = input("Enter your name: ")

print("Good Afternoon, " + name)

Enter your name: Tony
Good Afternoon, Tony


2. Improve Readability of a Message (Easy)

Concept: Escape sequences (\n, \t)

Question:
Format the following string using escape sequences to make it more readable with proper spacing and new lines.

`text = "Hello John, this course is amazing. Thank you!"`

In [ ]:
text = "Hello John,\n\tThis course is amazing.\nThank you!"
print(text)

Hello John,
	This course is amazing.
Thank you!


3. Write a program to check whether a given sentence contains two consecutive spaces.


In [ ]:
text = "Good evening  everyone, how are you"

print(text.find("  "))

12


4. Replace Extra Spaces with Single Space

Concept: replace()

Question:
Write a program to replace double spaces in a sentence with a single space.

In [ ]:
text = input("Enter a sentence: ")
text = text.replace("  ", " ")
print(text)

Enter a sentence: hello  everyone how are you
hello everyone how are you


5. Fill a Message Template

Concept: replace()

Question:
Given a message template with placeholders, replace them with user-provided values.

In [ ]:
message = '''
Hello <name>,
Welcome to the team!
Date: <date>
'''

name = input("Enter your name: ")
date = input("Enter date: ")

message = message.replace("<name>", name)
message = message.replace("<date>", date)

print(message)

Enter your name: Rohan
Enter date: 9 Apr

Hello Rohan,
Welcome to the team!
Date: 9 Apr



# Chapter 4 `LISTS AND TUPLES`

Python lists are containers to store a set of values of any data type.


## **LIST**

### 4.1 LIST INDEXING
A list can be indexed just like a string.

In [ ]:
l1 = ["red", 4, 9, "apple"]

In [ ]:
l1[0] # 7

'red'

In [ ]:
l1[1] # 9

4

In [ ]:
#l1[70] # error
l1[0:2] # [7,9] #list slicing

['red', 4]

In [ ]:
my_items = [10, "Banana", True, 20.5, "Cherry", "Date"]

print(my_items[0])
my_items[0] = 100 # Change the first element

print(my_items[0])
print(my_items[2:5]) # Slice from index 2 to 4 (exclusive of 5)

10
100
[True, 20.5, 'Cherry']


### 4.2 LIST METHODS.

Consider the following list:

In [ ]:
l1 = [1,8,7,2,21,15]

• l1.sort(): updates the list to [1,2,7,8,15,21]

• l1.reverse(): updates the list to [15,21,2,7,8,1]

• l1.append(8): adds 8 at the end of the list

• l1.insert(3,8): This will add 8 at 3 index

• l1.pop(2): Will delete element at index 2
and return its value.

• l1.remove(21): Will remove 21 from the list.

In [ ]:
l1.append(8)

In [ ]:
l1

[1, 8, 7, 2, 21, 15, 8]

In [ ]:
# Demonstration of list.sort()
l1 = [1, 8, 7, 2, 21, 15]
l1.sort() # Sorts the list in ascending order
print(f"List after sort(): {l1}")

List after sort(): [1, 2, 7, 8, 15, 21]


In [ ]:
# Demonstration of list.reverse()
l1 = [1, 8, 7, 2, 21, 15]
l1.reverse() # Reverses the order of elements
print(f"List after reverse(): {l1}")

List after reverse(): [15, 21, 2, 7, 8, 1]


In [ ]:
# Demonstration of list.insert(index, element)
l1 = [1, 8, 7, 2, 21, 15]
l1.insert(3, 88) # Inserts 88 at index 3
print(f"List after insert(3, 88): {l1}")

List after insert(3, 88): [1, 8, 7, 88, 2, 21, 15]


In [ ]:
# Demonstration of list.pop(index)
l1 = [1, 8, 7, 2, 21, 15]
popped_element = l1.pop(2) # Removes element at index 2 and returns it
print(f"Popped element: {popped_element}")
print(f"List after pop(2): {l1}")

Popped element: 7
List after pop(2): [1, 8, 2, 21, 15]


In [ ]:
# Demonstration of list.remove(value)
l1 = [1, 8, 7, 2, 21, 15, 8]
l1.remove(8) # Removes the first occurrence of 8
print(f"List after remove(8): {l1}")

List after remove(8): [1, 7, 2, 21, 15, 8]


### 4.2.1 Common List Operations

In [ ]:
# min(), max(), sum() for lists
numbers = [10, 5, 20, 15, 30]
print(f"List: {numbers}")
print(f"Minimum element: {min(numbers)}")
print(f"Maximum element: {max(numbers)}")
print(f"Sum of elements: {sum(numbers)}")

# List concatenation
list1 = [1, 2, 3]
list2 = [4, 5, 6]
combined_list = list1 + list2
print(f"Concatenated list: {combined_list}")

# List repetition
repeated_list = list1 * 2
print(f"Repeated list: {repeated_list}")

List: [10, 5, 20, 15, 30]
Minimum element: 5
Maximum element: 30
Sum of elements: 80
Concatenated list: [1, 2, 3, 4, 5, 6]
Repeated list: [1, 2, 3, 1, 2, 3]


## **Tuple**

###4.3 TUPLES IN PYTHON

A tuple is an immutable data type in python.

In [ ]:
a = () # empty tuple
a = (1,) # tuple with only one element needs a comma
a = (1,7,2) # tuple with more than one element

In [ ]:
a = (10, 20, 30, "Python", True, 3.14)
print(a)
print(type(a))

(10, 20, 30, 'Python', True, 3.14)
<class 'tuple'>


### 4.4 Tuple methods

In [ ]:
a = (100, 200, 300, 400, True, 250, "Python", "Java")
print(a)

(100, 200, 300, 400, True, 250, 'Python', 'Java')


In [ ]:
no = a.count(250)
print(no)

1


In [ ]:
i = a.index(400)
print(i)

3


In [ ]:
print(len(a))

8


### 4.4.1 Common Tuple Operations

In [ ]:
# min(), max(), sum() for tuples
numbers_tuple = (10, 5, 20, 15, 30)
print(f"Tuple: {numbers_tuple}")
print(f"Minimum element: {min(numbers_tuple)}")
print(f"Maximum element: {max(numbers_tuple)}")
print(f"Sum of elements: {sum(numbers_tuple)}")

# Tuple concatenation
tuple1 = (1, 2, 3)
tuple2 = (4, 5, 6)
combined_tuple = tuple1 + tuple2
print(f"Concatenated tuple: {combined_tuple}")

# Tuple repetition
repeated_tuple = tuple1 * 2
print(f"Repeated tuple: {repeated_tuple}")

Tuple: (10, 5, 20, 15, 30)
Minimum element: 5
Maximum element: 30
Sum of elements: 80
Concatenated tuple: (1, 2, 3, 4, 5, 6)
Repeated tuple: (1, 2, 3, 1, 2, 3)


### 4.5 Lists vs. Tuples: Key Differences

While both lists and tuples are used to store collections of items, they have fundamental differences that dictate their use cases:

*   **Mutability:**
    *   **Lists:** Are **mutable**, meaning their elements can be changed (added, removed, modified) after creation. This makes them suitable for collections of items that may need to be updated.
    *   **Tuples:** Are **immutable**, meaning their elements cannot be changed after creation. This makes them ideal for fixed collections of items or as dictionary keys (since keys must be immutable).

*   **Syntax:**
    *   **Lists:** Defined using square brackets `[]`.
    *   **Tuples:** Defined using parentheses `()`.

*   **Performance:**
    *   Tuples are generally slightly faster than lists for iteration and access because their size is fixed.

*   **Use Cases:**
    *   **Lists:** Use when you expect the collection to change (e.g., a list of users, a shopping cart).
    *   **Tuples:** Use when you need a fixed collection of items (e.g., coordinates, RGB color values, database records) or as arguments to functions where the data should not be altered.

## Questions

1. Store Car Brands

Concept: List, input()

Question:
Write a program to take names of 4 car brands from the user and store them in a list. Then display the list.

In [ ]:
car1 = input("Enter car brand 1: ")
car2 = input("Enter car brand 2: ")
car3 = input("Enter car brand 3: ")
car4 = input("Enter car brand 4: ")

cars = [car1, car2, car3, car4]

print("Car Brands:", cars)

Enter car brand 1: Audi
Enter car brand 2: BMW
Enter car brand 3: Kia
Enter car brand 4: Toyota
Car Brands: ['Audi', 'BMW', 'Kia', 'Toyota']


2. Arrange Car Prices

Concept: List, sort()

Question:
Take prices of 5 cars from the user and display them in increasing order.

In [ ]:
prices = []

p1 = int(input("Enter price 1: "))
prices.append(p1)
p2 = int(input("Enter price 2: "))
prices.append(p2)
p3 = int(input("Enter price 3: "))
prices.append(p3)
p4 = int(input("Enter price 4: "))
prices.append(p4)
p5 = int(input("Enter price 5: "))
prices.append(p5)

prices.sort()

print("Sorted Prices:", prices)

Enter price 1: 250
Enter price 2: 210
Enter price 3: 350
Enter price 4: 110
Enter price 5: 470
Sorted Prices: [110, 210, 250, 350, 470]


3. Try Updating a Car Collection (Tuple)

Concept: Tuple immutability

Question:
Create a tuple of 3 car brands and try to change one of them. Observe the result.

In [ ]:
cars = ("BMW", "Audi", "Tesla")

#cars[0] = "Mercedes"  # This will give an error

print(cars)

('BMW', 'Audi', 'Tesla')


4. Total Cost of Selected Cars

Concept: List, sum()

Question:
Store prices of 4 selected cars in a list and calculate the total cost.

In [ ]:
prices = [500000, 800000, 1200000, 700000]

total = sum(prices)

print("Total Cost:", total)

Total Cost: 3200000


5. Count a Specific Brand in Tuple

Concept: Tuple count()

Question:
Given a tuple of car brands, count how many times "Toyota" appears.

`cars = ("Toyota", "BMW", "Toyota", "Audi", "Toyota")`

In [ ]:
cars = ("Toyota", "BMW", "Toyota", "Audi", "Toyota")

count_toyota = cars.count("Toyota")

print("Toyota appears:", count_toyota, "times")

Toyota appears: 3 times


#CHAPTER 5 `DICTIONARY & SETS`


## **DICTIONARIES**

Dictionaries in Python are unordered collections of data values, used to store data values like a map, which, unlike other Data Types that hold only a single value as an element, holds `key:value` pairs. Each `key:value` pair in a Dictionary is separated by a comma, and the entire dictionary is enclosed in curly braces `{}`.

### 5.1 Creating and Accessing Dictionaries

Dictionaries are mutable, meaning they can be changed after they are created.

In [ ]:
# Creating an empty dictionary
my_dict = {}
print(f"Empty dictionary: {my_dict}")

# Creating a dictionary with key-value pairs
student = {
    'name': 'Alice',
    'age': 20,
    'major': 'Computer Science',
    'courses': ['Math', 'Physics', 'Programming']
}
print(f"Student dictionary: {student}")

# Accessing values using keys
print(f"Student's name: {student['name']}")
print(f"Student's age: {student.get('age')}") # Using .get() for safer access

# Trying to access a non-existent key with .get() (returns None by default)
print(f"Student's GPA (using get): {student.get('gpa')}")
# Accessing a non-existent key directly will raise a KeyError
# print(student['gpa'])

Empty dictionary: {}
Student dictionary: {'name': 'Alice', 'age': 20, 'major': 'Computer Science', 'courses': ['Math', 'Physics', 'Programming']}
Student's name: Alice
Student's age: 20
Student's GPA (using get): None


### 5.2 Modifying Dictionaries

You can add new key-value pairs or update existing ones.

In [ ]:
student = {
    'name': 'Alice',
    'age': 20,
    'major': 'Computer Science'
}

# Adding a new key-value pair
student['gpa'] = 3.8
print(f"After adding GPA: {student}")

# Updating an existing value
student['age'] = 21
print(f"After updating age: {student}")

# Using update() to add multiple items or update existing ones
student.update({
    'city': 'New York',
    'age': 22 # Updates age again
})
print(f"After update() method: {student}")

After adding GPA: {'name': 'Alice', 'age': 20, 'major': 'Computer Science', 'gpa': 3.8}
After updating age: {'name': 'Alice', 'age': 21, 'major': 'Computer Science', 'gpa': 3.8}
After update() method: {'name': 'Alice', 'age': 22, 'major': 'Computer Science', 'gpa': 3.8, 'city': 'New York'}


### 5.3 Removing Elements from Dictionaries

Elements can be removed using `del`, `pop()`, or `clear()`.

In [ ]:
student = {
    'name': 'Alice',
    'age': 22,
    'major': 'Computer Science',
    'gpa': 3.8,
    'city': 'New York'
}

# Removing a key-value pair using del
del student['city']
print(f"After deleting 'city': {student}")

# Removing a key-value pair using pop() (returns the value)
removed_gpa = student.pop('gpa')
print(f"Removed GPA: {removed_gpa}")
print(f"After pop('gpa'): {student}")

# Using popitem() to remove and return an arbitrary (last inserted in Python 3.7+) key-value pair
last_item = student.popitem()
print(f"Removed item with popitem(): {last_item}")
print(f"After popitem(): {student}")

# Clearing all items from the dictionary
student.clear()
print(f"After clear(): {student}")

After deleting 'city': {'name': 'Alice', 'age': 22, 'major': 'Computer Science', 'gpa': 3.8}
Removed GPA: 3.8
After pop('gpa'): {'name': 'Alice', 'age': 22, 'major': 'Computer Science'}
Removed item with popitem(): ('major', 'Computer Science')
After popitem(): {'name': 'Alice', 'age': 22}
After clear(): {}


### 5.4 Dictionary Methods

Important methods to retrieve keys, values, and items.

In [ ]:
student = {
    'name': 'Bob',
    'age': 25,
    'major': 'Physics',
    'gpa': 3.5
}

# Get all keys
keys = student.keys()
print(f"Keys: {keys}, Type: {type(keys)}")

# Get all values
values = student.values()
print(f"Values: {values}, Type: {type(values)}")

# Get all items (key-value pairs)
items = student.items()
print(f"Items: {items}, Type: {type(items)}")

# Iterating through a dictionary
print("\nIterating through items:")
for key, value in student.items():
    print(f"{key}: {value}")

Keys: dict_keys(['name', 'age', 'major', 'gpa']), Type: <class 'dict_keys'>
Values: dict_values(['Bob', 25, 'Physics', 3.5]), Type: <class 'dict_values'>
Items: dict_items([('name', 'Bob'), ('age', 25), ('major', 'Physics'), ('gpa', 3.5)]), Type: <class 'dict_items'>

Iterating through items:
name: Bob
age: 25
major: Physics
gpa: 3.5


### 5.5 Dictionary Comprehensions

Similar to list comprehensions, dictionary comprehensions provide a concise way to create dictionaries.

In [ ]:
# Create a dictionary of squares
squares = {num: num**2 for num in range(5)}
print(f"Squares dictionary: {squares}")

# Another simple example: Create a dictionary where keys are fruits and values are their length

fruits = ['apple', 'banana', 'cherry']
fruit_lengths = {fruit: len(fruit) for fruit in fruits}
print(f"Fruit lengths dictionary: {fruit_lengths}")

# Create a dictionary from two lists
keys = ['a', 'b', 'c']
values = [1, 2, 3]
my_dict_from_lists = {k: v for k, v in zip(keys, values)}
print(f"Dictionary from lists: {my_dict_from_lists}")

Squares dictionary: {0: 0, 1: 1, 2: 4, 3: 9, 4: 16}
Fruit lengths dictionary: {'apple': 5, 'banana': 6, 'cherry': 6}
Dictionary from lists: {'a': 1, 'b': 2, 'c': 3}


## **SETS**

Sets are unordered collections of unique elements. They are mutable, but the elements within a set must be immutable (like numbers, strings, tuples).

### 5.6 Creating and Modifying Sets

Sets are useful for mathematical set operations like union, intersection, and difference, and for efficiently checking for membership.

In [ ]:
# Creating an empty set
empty_set = set() # Note: {} creates an empty dictionary, not an empty set
print(f"Empty set: {empty_set}")

# Creating a set with elements (duplicates are automatically removed)
my_set = {1, 2, 3, 2, 4, 1}
print(f"Set with unique elements: {my_set}")

# Creating a set from a list
list_to_set = set([5, 6, 7, 6])
print(f"Set from a list: {list_to_set}")

# Adding elements to a set
my_set.add(5)
print(f"After adding 5: {my_set}")
my_set.add(1) # Adding an existing element does nothing
print(f"After adding 1 again: {my_set}")

# Adding multiple elements using update()
my_set.update([6, 7, 8])
print(f"After updating with a list: {my_set}")

Empty set: set()
Set with unique elements: {1, 2, 3, 4}
Set from a list: {5, 6, 7}
After adding 5: {1, 2, 3, 4, 5}
After adding 1 again: {1, 2, 3, 4, 5}
After updating with a list: {1, 2, 3, 4, 5, 6, 7, 8}


### 5.7 Removing Elements from Sets

Elements can be removed using `remove()`, `discard()`, `pop()`, or `clear()`.

In [ ]:
my_set = {1, 2, 3, 4, 5, 6}

# Removing an element using remove()
my_set.remove(3)
print(f"After removing 3: {my_set}")
# my_set.remove(10) # This would raise a KeyError if element not found

# Removing an element using discard()
my_set.discard(2)
print(f"After discarding 2: {my_set}")
my_set.discard(10) # Discarding a non-existent element does nothing
print(f"After discarding 10 (non-existent): {my_set}")

# Removing an arbitrary element using pop()
popped_element = my_set.pop()
print(f"Popped element: {popped_element}")
print(f"After pop(): {my_set}")

# Clearing all elements from the set
my_set.clear()
print(f"After clear(): {my_set}")

After removing 3: {1, 2, 4, 5, 6}
After discarding 2: {1, 4, 5, 6}
After discarding 10 (non-existent): {1, 4, 5, 6}
Popped element: 1
After pop(): {4, 5, 6}
After clear(): set()


### 5.8 Set Operations

Sets support mathematical operations like union, intersection, difference, and symmetric difference.

In [ ]:
set_a = {1, 2, 3, 4, 5}
set_b = {4, 5, 6, 7, 8}

print(f"Set A: {set_a}")
print(f"Set B: {set_b}")

# Union: All unique elements from both sets
union_set = set_a.union(set_b) # or set_a | set_b
print(f"Union (A | B): {union_set}")

# Intersection: Common elements in both sets
intersection_set = set_a.intersection(set_b) # or set_a & set_b
print(f"Intersection (A & B): {intersection_set}")

# Difference: Elements in A but not in B
difference_set_ab = set_a.difference(set_b) # or set_a - set_b
print(f"Difference (A - B): {difference_set_ab}")

# Difference: Elements in B but not in A
difference_set_ba = set_b.difference(set_a) # or set_b - set_a
print(f"Difference (B - A): {difference_set_ba}")

# Symmetric Difference: Elements in either A or B, but not both
symmetric_difference_set = set_a.symmetric_difference(set_b) # or set_a ^ set_b
print(f"Symmetric Difference (A ^ B): {symmetric_difference_set}")

# Subset and Superset
set_c = {1, 2}
print(f"Is {set_c} a subset of {set_a}? {set_c.issubset(set_a)}")
print(f"Is {set_a} a superset of {set_c}? {set_a.issuperset(set_c)}")

Set A: {1, 2, 3, 4, 5}
Set B: {4, 5, 6, 7, 8}
Union (A | B): {1, 2, 3, 4, 5, 6, 7, 8}
Intersection (A & B): {4, 5}
Difference (A - B): {1, 2, 3}
Difference (B - A): {8, 6, 7}
Symmetric Difference (A ^ B): {1, 2, 3, 6, 7, 8}
Is {1, 2} a subset of {1, 2, 3, 4, 5}? True
Is {1, 2, 3, 4, 5} a superset of {1, 2}? True


##Questions

1.  **Product Details Dictionary**

    Create a dictionary called `product_details` to store information about a product. Include its `name` (e.g., "Wireless Mouse"), `price` (e.g., 25.99), and `in_stock` status (e.g., `True`). Then, print the product's name and its price.

In [ ]:
product_details = {
    "name": "Wireless Mouse",
    "price": 25.99,
    "in_stock": True
}

print(f"Product Name: {product_details['name']}")
print(f"Product Price: ${product_details['price']}")

Product Name: Wireless Mouse
Product Price: $25.99


2.  **Update Student Grades**

    You have a dictionary representing a student's grades: `student_grades = {'Math': 90, 'Science': 85}`. Add a new subject 'History' with a grade of 92. Then, update the 'Science' grade to 88. Print the final dictionary.

In [ ]:
student_grades = {'Math': 90, 'Science': 85}

# Add a new subject
student_grades['History'] = 92

# Update an existing grade
student_grades['Science'] = 88

print("Updated student grades:", student_grades)

Updated student grades: {'Math': 90, 'Science': 88, 'History': 92}


3.  **Find Unique Fruits**

    You are given a list of favorite fruits, but some fruits are repeated: `favorite_fruits = ['apple', 'banana', 'orange', 'apple', 'grape', 'banana']`. Convert this list into a set to find out all the unique fruits, and then print the set.

In [ ]:
favorite_fruits = ['apple', 'banana', 'orange', 'apple', 'grape', 'banana']

unique_fruits = set(favorite_fruits)

print("Unique favorite fruits:", unique_fruits)

Unique favorite fruits: {'grape', 'apple', 'banana', 'orange'}


4.  **Common Favorite Colors**

    Two friends, Alice and Bob, listed their favorite colors. Alice's favorites are `{'red', 'blue', 'green'}` and Bob's favorites are `{'green', 'yellow', 'blue'}`. Find out which colors they both like using a set operation.

In [ ]:
alice_colors = {'red', 'blue', 'green'}
bob_colors = {'green', 'yellow', 'blue'}

common_colors = alice_colors.intersection(bob_colors) # Or alice_colors & bob_colors

print("Colors both Alice and Bob like:", common_colors)

Colors both Alice and Bob like: {'green', 'blue'}


5.  **Empty Collection Type**

    What data type will `my_empty_collection = {}` create in Python? How would you create an empty set instead? Print the type of both variables.

In [ ]:
my_empty_collection = {}
print(f"Type of {{}}: {type(my_empty_collection)}")

# To create an empty set:
my_empty_set = set()
print(f"Type of empty set(): {type(my_empty_set)}")

Type of {}: <class 'dict'>
Type of empty set(): <class 'set'>


6.  **Adding a List to a Set**

    Can you add a list `[1, 2, 3]` to a set `my_set = {10, 20}`? Explain why or why not, and demonstrate with code what happens if you try.

In [ ]:
my_set = {10, 20}
my_list = [1, 2, 3]

print(f"Initial set: {my_set}")

try:
    my_set.add(my_list)
    print("Set after attempting to add list:", my_set)
except TypeError as e:
    print(f"Error: {e}")
    print("\nExplanation: You cannot add a list to a set because lists are mutable (changeable). Set elements must be immutable (hashable), meaning their value cannot change after they are created. Tuples, numbers, and strings are examples of immutable types that can be added to a set.")

Initial set: {10, 20}
Error: unhashable type: 'list'

Explanation: You cannot add a list to a set because lists are mutable (changeable). Set elements must be immutable (hashable), meaning their value cannot change after they are created. Tuples, numbers, and strings are examples of immutable types that can be added to a set.


#CHAPTER 6 `CONDITIONAL EXPRESSION`

Conditional expressions, or `if` statements, are fundamental in programming for controlling the flow of execution based on certain conditions. They allow your program to make decisions and execute different blocks of code depending on whether a condition is `True` or `False`.

### 6.1 The `if` Statement

The `if` statement is the simplest form of a conditional. It executes a block of code only if a specified condition is `True`.

In [ ]:
score = 75

if score >= 70:
    print("Congratulations! You passed.")

print("End of program.")

score = 60
if score >= 70:
    print("Congratulations! You passed.")

print("End of program (second check).")

Congratulations! You passed.
End of program.
End of program (second check).


### 6.2 The `if-else` Statement

The `if-else` statement allows you to specify a block of code to be executed if the condition is `True`, and an alternative block to be executed if the condition is `False`.

In [ ]:
age = 18

if age >= 18:
    print("You are eligible to vote.")
else:
    print("You are not yet eligible to vote.")

price = 100
discount_applied = False

if price > 50:
    discount_amount = price * 0.10
    final_price = price - discount_amount
    discount_applied = True
    print(f"Discount applied! Final price: ${final_price:.2f}")
else:
    final_price = price
    print(f"No discount. Final price: ${final_price:.2f}")

print(f"Was discount applied? {discount_applied}")

You are eligible to vote.
Discount applied! Final price: $90.00
Was discount applied? True


### 6.3 The `if-elif-else` Statement

When you have multiple conditions to check, you can use the `if-elif-else` chain. The `elif` (short for 'else if') keyword allows you to check multiple expressions for `True` and execute a block of code as soon as one of the conditions evaluates to `True`.

In [ ]:
temperature = 25

if temperature > 30:
    print("It's a hot day!")
elif temperature > 20:
    print("It's a pleasant day.")
elif temperature > 10:
    print("It's a bit cool.")
else:
    print("It's cold!")

grade = 88

if grade >= 90:
    print("Grade: A")
elif grade >= 80:
    print("Grade: B")
elif grade >= 70:
    print("Grade: C")
elif grade >= 60:
    print("Grade: D")
else:
    print("Grade: F")

It's a pleasant day.
Grade: B


### 6.4 Nested `if` Statements

You can have `if` statements inside other `if` statements. This is called nesting. It's useful when you need to check a secondary condition only if a primary condition is met.

In [ ]:
num = 15

if num > 0:
    print("Number is positive.")
    if num % 2 == 0:
        print("Number is also even.")
    else:
        print("Number is also odd.")
else:
    print("Number is not positive (it's zero or negative).")


is_admin = True
is_logged_in = True

if is_logged_in:
    print("User is logged in.")
    if is_admin:
        print("Welcome, Admin!")
    else:
        print("Welcome, User!")
else:
    print("Please log in.")

Number is positive.
Number is also odd.
User is logged in.
Welcome, Admin!


### 6.5 Shorthand `if` (Ternary Operator)

For simple `if-else` statements, Python offers a concise one-line conditional expression, also known as the ternary operator. It makes your code more compact for assignments based on a condition.

In [ ]:
a = 10
b = 5

# Traditional if-else
if a > b:
    max_val = a
else:
    max_val = b
print(f"Traditional max_val: {max_val}")

# Shorthand if-else
max_val_short = a if a > b else b
print(f"Shorthand max_val: {max_val_short}")


weather = "rainy"
status = "Bring an umbrella" if weather == "rainy" else "Enjoy the day"
print(f"Weather status: {status}")

Traditional max_val: 10
Shorthand max_val: 10
Weather status: Bring an umbrella


### 6.6 Logical Operators (`and`, `or`, `not`) with `if`

Logical operators are used to combine conditional statements. They allow you to create more complex conditions.

In [ ]:
age = 25
has_license = True

# 'and' operator: Both conditions must be True
if age >= 18 and has_license:
    print("You can drive.")
else:
    print("You cannot drive.")


# 'or' operator: At least one condition must be True
is_weekend = False
has_day_off = True

if is_weekend or has_day_off:
    print("It's a leisure day!")
else:
    print("Time to work.")


# 'not' operator: Inverts the boolean value
is_hungry = False

if not is_hungry:
    print("I'm not hungry right now.")
else:
    print("I'm hungry!")


# Combining logical operators
has_ticket = True
is_vip = False
entry_allowed = False

if has_ticket and (is_vip or age >= 21): # age is 25 from above
    entry_allowed = True
    print("Entry granted!")
else:
    print("Entry denied.")

print(f"Final entry status: {entry_allowed}")

You can drive.
It's a leisure day!
I'm not hungry right now.
Entry granted!
Final entry status: True


### 6.7 The `pass` Statement

The `pass` statement is a null operation; nothing happens when it executes. It's useful as a placeholder where a statement is syntactically required but you don't want any code to execute yet. This avoids syntax errors in blocks that are intentionally left empty.

In [ ]:
hour = 10

if hour < 12:
    # TODO: Implement morning greeting logic later
    pass
else:
    print("Good afternoon!")

print("Program continues...")

Program continues...


##Questions

1.  **Check Even or Odd**

    Write a program that takes an integer input from the user and prints whether the number is 'Even' or 'Odd'.

In [ ]:
num = int(input("Enter an integer: "))

if num % 2 == 0:
    print("Even")
else:
    print("Odd")

Enter an integer: 7
Odd


2.  **Determine Voting Eligibility**

    Ask the user for their age. If they are 18 or older, print "You are eligible to vote."; otherwise, print "You are not eligible to vote yet."

In [ ]:
age = int(input("Enter your age: "))

if age >= 18:
    print("You are eligible to vote.")
else:
    print("You are not eligible to vote yet.")

Enter your age: 22
You are eligible to vote.


3.  **Classify a Triangle**

    Given three side lengths of a triangle `s1`, `s2`, `s3`, determine if it is an 'Equilateral', 'Isosceles', or 'Scalene' triangle.

In [ ]:
s1 = 5
s2 = 5
s3 = 5

if s1 == s2 and s2 == s3:
    print("Equilateral Triangle")
elif s1 == s2 or s1 == s3 or s2 == s3:
    print("Isosceles Triangle")
else:
    print("Scalene Triangle")

Equilateral Triangle


4.  **Check for Leap Year**

    Write a program that takes a year as input and checks if it's a leap year. (A leap year is divisible by 4, but not by 100 unless it is also divisible by 400).

In [ ]:
year = int(input("Enter a year: "))

if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
    print(f"{year} is a leap year.")
else:
    print(f"{year} is not a leap year.")

Enter a year: 2024
2024 is a leap year.


5.  **Traffic Light Status**

    Simulate a traffic light. If the color is 'red', print "Stop". If it's 'yellow', print "Prepare to stop". If it's 'green', print "Go". For any other input, print "Invalid color."

In [ ]:
color = input("Enter traffic light color (red, yellow, green): ").lower()

if color == 'red':
    print("Stop")
elif color == 'yellow':
    print("Prepare to stop")
elif color == 'green':
    print("Go")
else:
    print("Invalid color.")

Enter traffic light color (red, yellow, green): red
Stop


6.  **Login System Check (Nested If)**

    Create a simple login system. Ask for a `username` and `password`. If the `username` is 'admin' AND the `password` is 'password123', print "Login Successful - Admin Access". If the username is 'user' AND the password is 'guest', print "Login Successful - User Access". Otherwise, print "Invalid Credentials".

In [ ]:
username = input("Enter username: ")
password = input("Enter password: ")

if username == 'admin':
    if password == 'password123':
        print("Login Successful - Admin Access")
    else:
        print("Invalid Credentials")
elif username == 'user':
    if password == 'guest':
        print("Login Successful - User Access")
    else:
        print("Invalid Credentials")
else:
    print("Invalid Credentials")

Enter username: user
Enter password: guest
Login Successful - User Access


# CHAPTER 7 `LOOPS`

### 7.1 Introduction to Loops
Loops are used to repeat a block of code multiple times. Instead of writing the same code over and over, we use loops to automate repetitive tasks, such as printing numbers from 1 to 1000 or processing items in a list.

### 7.2 The `while` Loop
A `while` loop keeps executing as long as a certain condition remains `True`.

**Syntax:**
```python
while condition:
    # body of loop
```

In [ ]:
# Example: Printing numbers from 1 to 5
i = 1
while i <= 5:
    print(f"Iteration {i}")
    i += 1  # Incrementing i is crucial to eventually stop the loop

print("Loop finished!")

Iteration 1
Iteration 2
Iteration 3
Iteration 4
Iteration 5
Loop finished!


while loop with list

In [ ]:
l = [1, "Tony", False, "This", 7, "Cherry", "Karan"]

i = 0

while(i<len(l)):
    print(l[i])
    i +=1

1
Tony
False
This
7
Cherry
Karan


### 7.3 The `for` Loop
A `for` loop is used to iterate over a sequence (like a list, tuple, or string). It automatically moves to the next item in the sequence until it reaches the end.

In [ ]:
for i in range(7):
  print(i)

0
1
2
3
4
5
6


In [ ]:
fruits = ["Apple", "Banana", "Cherry"]

print("My fruit list:")
for fruit in fruits:
    print(f"- {fruit}")

My fruit list:
- Apple
- Banana
- Cherry


In [ ]:
name = "Tom Cruise"

for i in name:
  print(i)

T
o
m
 
C
r
u
i
s
e


### 7.4 The `range()` Function
The `range()` function generates a sequence of numbers. It is commonly used with `for` loops to run code a specific number of times.

- `range(stop)`: 0 to stop-1
- `range(start, stop)`: start to stop-1
- `range(start, stop, step)`: start to stop-1, jumping by 'step'

In [ ]:
print("Counting by twos up to 10:")
for num in range(2, 11, 2):
    print(num)

Counting by twos up to 10:
2
4
6
8
10


### 7.5 Loop Control Statements
Sometimes you need to change the behavior of a loop while it is running:
- **`break`**: Exits the loop immediately.
- **`continue`**: Skips the current iteration and moves to the next one.
- **`pass`**: A placeholder that does nothing (prevents errors in empty blocks).

In [ ]:
print("--- Testing break (stop at 5) ---")
for i in range(1, 10):
    if i == 5:
        break
    print(i)

print("\n--- Testing continue (skip even numbers) ---")
for i in range(1, 6):
    if i % 2 == 0:
        continue
    print(i)

print("\n--- Testing pass ---")
for i in range(3):
    pass # I will add logic here later
print("Pass executed without error.")

--- Testing break (stop at 5) ---
1
2
3
4

--- Testing continue (skip even numbers) ---
1
3
5

--- Testing pass ---
Pass executed without error.


### 7.6 For-Else Loop
Python allows an `else` block after a loop. The `else` code runs only if the loop finished naturally (i.e., it was NOT stopped by a `break`).

In [ ]:
numbers = [1, 3, 5]

for n in numbers:
    print(n)
else:
    print("Loop completed successfully (no break encountered)!")

1
3
5
Loop completed successfully (no break encountered)!


## Practice Questions

1. **Multiplication Table**

Write a program to print the multiplication table of a given number entered by the user using a `for` loop.

In [ ]:
num = int(input("Enter a number: "))

for i in range(1, 11):
    print(f"{num} x {i} = {num * i}")

Enter a number: 7
7 x 1 = 7
7 x 2 = 14
7 x 3 = 21
7 x 4 = 28
7 x 5 = 35
7 x 6 = 42
7 x 7 = 49
7 x 8 = 56
7 x 9 = 63
7 x 10 = 70


2. **Sum of Natural Numbers**

Write a program to find the sum of first `n` natural numbers using a `while` loop.

In [ ]:
n = int(input("Enter value for n: "))
sum_numbers = 0
i = 1

while i <= n:
    sum_numbers += i
    i += 1

print(f"The sum of first {n} numbers is: {sum_numbers}")

Enter value for n: 5
The sum of first 5 numbers is: 15


3. **Greeting List Members**

Given a list of names, use a `for` loop to greet only those whose names start with 'S'.

`names = ["Sahil", "Rohan", "Sameer", "Rahul"]`

In [ ]:
names = ["Sahil", "Rohan", "Sameer", "Rahul"]

for name in names:
    if name.startswith("S"):
        print(f"Hello {name}!")

Hello Sahil!
Hello Sameer!


4. **Factorial Calculation**

Write a program to calculate the factorial of a given number using a `for` loop.

In [ ]:
num = int(input("Enter a number: "))
factorial = 1

for i in range(1, num + 1):
    factorial *= i

print(f"The factorial of {num} is {factorial}")

Enter a number: 5
The factorial of 5 is 120


5. **Prime Number Check**

Write a program to find whether a given number is prime or not.

In [ ]:
num = int(input("Enter a number: "))
is_prime = True

if num < 2:
    is_prime = False
else:
    for i in range(2, num):
        if num % i == 0:
            is_prime = False
            break

if is_prime:
    print(f"{num} is a prime number.")
else:
    print(f"{num} is not a prime number.")

Enter a number: 7
7 is a prime number.


6. **Pattern Printing**

Write a program to print the following pattern for a given `n` (e.g., n=3):

```
*
**
***
```

In [ ]:
n = int(input("Enter number of rows: "))

for i in range(1, n + 1):
    print("*" * i)

Enter number of rows: 5
*
**
***
****
*****


In [ ]:
print("--- Pyramid Pattern ---")
n = 5
for i in range(1, n + 1):
    # Print spaces
    print(" " * (n - i), end="")
    # Print stars
    print("*" * (2 * i - 1))

--- Pyramid Pattern ---
    *
   ***
  *****
 *******
*********


In [ ]:
print("\n--- Diamond Pattern ---")
# Upper part
for i in range(1, n + 1):
    print(" " * (n - i) + "*" * (2 * i - 1))
# Lower part
for i in range(n - 1, 0, -1):
    print(" " * (n - i) + "*" * (2 * i - 1))


--- Diamond Pattern ---
    *
   ***
  *****
 *******
*********
 *******
  *****
   ***
    *


# Chapter 8 `FUNCTIONS & RECURSIONS`

### 8.1 Introduction to Functions
A function is a reusable block of code that performs a specific task. Instead of writing the same logic multiple times, you define it once in a function and 'call' it whenever needed.

**Syntax:**
```python
def function_name(parameters):
    # Body of the function
    return value
```

In [ ]:
def greet(name):
    """This function greets the person passed in as a parameter"""
    print(f"Hello, {name}! Good morning.")

# Calling the function
greet("Guddu")
greet("Rohan")

Hello, Guddu! Good morning.
Hello, Rohan! Good morning.


### 8.2 Types of Functions in Python
1. **Built-in Functions:** Functions already provided by Python (e.g., `print()`, `len()`, `range()`).
2. **User-defined Functions:** Functions created by the programmer to perform specific tasks.

### 8.3 Functions with Arguments and Return Values
Functions can take inputs (arguments) and give back a result (return value).

In [ ]:
def sum_numbers(a, b):
    return a + b

result = sum_numbers(10, 20)
print(f"The sum is: {result}")

The sum is: 30


### 8.4 Default Parameter Values
You can provide a default value to a parameter. If no argument is provided during the function call, the default value is used.

In [ ]:
# def power(base, exponent=2):
#     return base ** exponent

# print(f"Power with default (5^2): {power(5)}")
# print(f"Power with custom argument (2^3): {power(2, 3)}")

def greet(name = "sameer"):
  print(f"hello, {name}")
greet() # name will be "sameer" in function body (default)
greet("Tom") # name will be "Tom" in function body (passed)


hello, sameer
hello, Tom


### 8.5 Introduction to Recursion
Recursion is a process where a function calls itself. It is often used to solve mathematical problems like factorials or Fibonacci sequences.

**Crucial Note:** A recursive function must have a **Base Case** to stop it from calling itself infinitely.

In [ ]:
def factorial_recursive(n):
    # Base case: factorial of 0 or 1 is 1
    if n == 0 or n == 1:
        return 1
    # Recursive case
    else:
        return n * factorial_recursive(n - 1)

num = 5
print(f"Factorial of {num} using recursion is: {factorial_recursive(num)}")

Factorial of 5 using recursion is: 120


### 8.6 How Recursion Works (Step-by-Step)

When you call `factorial_recursive(5)`, the function doesn't return a result immediately. It keeps calling itself until it hits the **base case**. Here is the mental model of the process:

**1. The Expansion Phase (Going Down):**
- `factorial_recursive(5)` returns `5 * factorial_recursive(4)`
- `factorial_recursive(4)` returns `4 * factorial_recursive(3)`
- `factorial_recursive(3)` returns `3 * factorial_recursive(2)`
- `factorial_recursive(2)` returns `2 * factorial_recursive(1)`
- `factorial_recursive(1)` returns `1` **(Base Case reached!)**

**2. The Substitution Phase (Coming Up):**
- Now the values start returning back up the chain:
- `factorial_recursive(1) = 1`
- `factorial_recursive(2) = 2 * 1 = 2`
- `factorial_recursive(3) = 3 * 2 = 6`
- `factorial_recursive(4) = 4 * 6 = 24`
- `factorial_recursive(5) = 5 * 24 = 120`

## Practice Questions

1. **Find the Maximum**

Write a function called `find_max(a, b, c)` that takes three numbers as arguments and returns the largest of them using conditional statements.

In [ ]:
def find_max(a, b, c):
    if (a >= b) and (a >= c):
        return a
    elif (b >= a) and (b >= c):
        return b
    else:
        return c

# Test the function
print(f"The maximum of 10, 25, and 15 is: {find_max(10, 25, 15)}")

The maximum of 10, 25, and 15 is: 25


2. **Celsius to Fahrenheit Converter**

Write a function to convert temperature from Celsius to Fahrenheit.
Formula: `F = (C * 9/5) + 32`

In [ ]:
def celsius_to_fahrenheit(c):
    return (c * 9/5) + 32

c_temp = 37
print(f"{c_temp}°C is equal to {celsius_to_fahrenheit(c_temp)}°F")

37°C is equal to 98.6°F


3. **Recursive Sum of Natural Numbers**

Write a recursive function to calculate the sum of the first `n` natural numbers. (e.g., if n=3, sum = 3+2+1=6)

In [ ]:
def recursive_sum(n):
    if n <= 1:
        return n
    return n + recursive_sum(n - 1)

num = 10
print(f"The sum of first {num} natural numbers is: {recursive_sum(num)}")

The sum of first 10 natural numbers is: 55


4. **Inverted Triangle Pattern**

Write a function `print_pattern(n)` that prints the first `n` lines of the following pattern:
```
***
**
*
```

In [ ]:
def print_pattern(n):
    for i in range(n, 0, -1):
        print("*" * i)

print_pattern(3)

***
**
*


5. **Inches to Centimeters**

Write a function that converts inches to centimeters (Note: 1 inch = 2.54 cm).

In [ ]:
def inch_to_cm(inch):
    return inch * 2.54

inches = 10
print(f"{inches} inches is {inch_to_cm(inches)} cm")

10 inches is 25.4 cm


6. **Word Remover and Stripper**

Write a function that removes a specific word from a string and strips any leading/trailing spaces from the result.

In [ ]:
def remove_and_strip(string, word):
    new_str = string.replace(word, "")
    return new_str.strip()

text = "   Python is fun   "
print(f"Modified string: '{remove_and_strip(text, 'fun')}'")

Modified string: 'Python is'


# Chapter 9 `FILE I/O`

Files are used to store data permanently. In Python, we use the `open()` function to handle files.

### 9.1 Opening a File and Modes

When opening a file, we specify a **mode**:
- `'r'`: **Read mode** (default). Opens a file for reading; error if the file doesn't exist.
- `'w'`: **Write mode**. Opens a file for writing; creates it if it doesn't exist or **overwrites** it if it does.
- `'a'`: **Append mode**. Opens a file for appending data at the end without deleting existing content.
- `'rb'` / `'wb'`: Used for **Binary files** (like images or PDFs).

In [ ]:
# 1. Writing to a file
f = open("myfile.txt", "w")
f.write("Hello, this is Chapter 9.\n")
f.write("Learning File I/O is essential.")

# 2. Closing the file
f.close()

print("File created and closed successfully.")

File created and closed successfully.


### ⚠️ CRITICAL: The Behavior of Write Mode (`'w'`)

It is vital to understand that when you open a file in `'w'` mode, Python **immediately wipes out** everything that was previously in that file before you even write a single line.

- **'w' (Write):** Deletes existing content and starts fresh.
- **'a' (Append):** Keeps existing content and adds new data to the end of the file.

Always double-check your mode before running your code!

### 9.2 Why do we use `f.close()`?

It is crucial to close a file after you are done with it because:
1. **Resource Management:** Operating systems have a limit on how many files can be open at once.
2. **Data Integrity:** Closing a file ensures that all buffered data is physically written to the disk.
3. **Prevents Corruption:** Leaving files open can lead to data loss if the program crashes.

### 9.3 Reading a File

We can read the entire content using `.read()`, a single line using `.readline()`, or all lines into a list using `.readlines()`.

In [ ]:
f = open("myfile.txt", "r")
content = f.read()
print("--- File Content ---")
print(content)
f.close()

--- File Content ---
Hello, this is Chapter 9.
Learning File I/O is essential.


### 9.4 The `with` Statement (Best Practice)

The `with` statement is the most optimized and safest way to handle files.

**Why use it?**
- It automatically closes the file for you as soon as the block of code is finished.
- It handles exceptions internally—even if an error occurs inside the block, the file is guaranteed to close.

In [ ]:
# No need to call f.close() manually here!
with open("myfile.txt", "r") as f:
    for line in f:
        print(f"Reading line: {line.strip()}")

Reading line: Hello, this is Chapter 9.
Reading line: Learning File I/O is essential.


## Practice Questions

1. **Log Entry Checker**

Write a program to read a file named `server_log.txt` and check if it contains the word 'Error' (case-insensitive). If it does, print a warning message.

In [ ]:
# Q1: Search in Server Logs
# Let's create the log file first
with open("server_log.txt", "w") as f:
    f.write("System Boot... OK\nDatabase Connection... OK\nError: Connection Timeout\nRetrying...")

with open("server_log.txt", "r") as f:
    content = f.read()

if "error" in content.lower():
    print("⚠️ ALERT: Critical Error found in logs!")
else:
    print("✅ System Status: Healthy")

⚠️ ALERT: Critical Error found in logs!


2. **Game High Score Updater**

Create a function `update_highscore(new_score)`. It should read `highscore.txt`. If the `new_score` is higher than the current high score, it should overwrite the file with the new value.

In [ ]:
import os
import builtins

def update_highscore(new_score):
    # Initialize current_high to 0
    current_high = 0

    # Check if the file exists before reading
    if os.path.exists("highscore.txt"):
        with open("highscore.txt", "r") as f:
            data = f.read()
            if data:
                current_high = int(data)

    if new_score > current_high:
        print(f"New High Score! {new_score}")
        with open("highscore.txt", "w") as f:
            f.write(builtins.str(new_score)) # Use builtins.str to avoid shadowing issues
    else:
        print(f"Try again! Current high score is {current_high}")

# Simulate a game session
update_highscore(85)

Try again! Current high score is 85


3. **Word Censorship Tool**

Write a program that reads `chat_history.txt` and replaces a list of prohibited words (e.g., 'donkey', 'idiot') with '#####'.

In [ ]:
# Q3: Censorship
prohibited_words = ["donkey", "idiot"]

with open("chat_history.txt", "w") as f:
    f.write("The user called the bot a donkey, but the bot was not an idiot.")

with open("chat_history.txt", "r") as f:
    chat_content = f.read()

for word in prohibited_words:
    chat_content = chat_content.replace(word, "#####")

with open("chat_history.txt", "w") as f:
    f.write(chat_content)

print("Censored Chat:")
print(chat_content)

Censored Chat:
The user called the bot a #####, but the bot was not an #####.


4. **Log Cleaner (Emptying File)**

Write a program that completely wipes out the content of a file named `temp_logs.txt` without deleting the file itself.

In [ ]:
# Q4: Clearing a file
with open("temp_logs.txt", "w") as f:
    f.write("Old logs that take up too much space...")

# Wiping content using 'w' mode
with open("temp_logs.txt", "w") as f:
    pass

print("temp_logs.txt has been cleared.")

temp_logs.txt has been cleared.


5. **Automated Backup**

Write a program to copy the entire contents of a file `data.txt` into another file `data_backup.txt`.

In [ ]:
# Q5: Backup Script
with open("data.txt", "w") as f:
    f.write("CRITICAL_USER_DATA_001: 12345")

with open("data.txt", "r") as f:
    content = f.read()

with open("data_backup.txt", "w") as f:
    f.write(content)

print("Backup completed successfully.")

Backup completed successfully.


6. **Line Count Analysis**

Write a program to count how many lines are present in a file named `notes.txt`.

In [ ]:
# Q6: Counting lines
with open("notes.txt", "w") as f:
    f.write("Line 1: Buy milk\nLine 2: Finish homework\nLine 3: Go to gym")

with open("notes.txt", "r") as f:
    lines = f.readlines()

print(f"Total lines in notes.txt: {len(lines)}")

Total lines in notes.txt: 3


In [ ]:
print(l)

[1, 'Tony', False, 'This', 7, 'Cherry', 'Karan']


# Chapter 10 `OBJECT ORIENTED PROGRAMMING`

### 10.1 Introduction to OOP
Solving a problem by creating objects is one of the most popular approaches in programming. This is called Object-Oriented Programming.

**Dry Principle:** OOP helps us follow the 'Don't Repeat Yourself' principle. Instead of writing separate variables and functions for every user or item, we create a template (Class) and reuse it.

### 10.2 Class and Object
- **Class:** A blueprint for creating objects (e.g., a blank Application Form).
- **Object:** An instantiation of a class (e.g., a form filled with 'John's' details).

In [ ]:
class Employee:
    language = "Python" # Class Attribute
    salary = 1200000

emp1 = Employee()
emp1.name = "John"
print(emp1.name, emp1.language, emp1.salary)

John Python 1200000


### 10.3 The `self` Parameter
`self` refers to the instance of the class. It is automatically passed with any function call from an object. It allows the function to access attributes and methods of that specific object.

In [ ]:
class Employee:
    language = "Python"
    salary = 1200000

    def getDetails(self):
        print(f"The language is {self.language}. The salary is {self.salary}")

emp1 = Employee()
emp1.getDetails() # This is equivalent to Employee.getDetails(emp1)

The language is Python. The salary is 1200000


### 10.4 The `__init__` Constructor
`__init__()` is a special method which is run as soon as an object is created. It is used to initialize the attributes of an object.

In [ ]:
class Employee:
    def __init__(self, name, salary, language):
        self.name = name
        self.salary = salary
        self.language = language
        print("I am creating an object")

rohan = Employee("Rohan", 1300000, "JavaScript")
print(rohan.name, rohan.salary, rohan.language)

I am creating an object
Rohan 1300000 JavaScript


### 10.5 Static Methods (`@staticmethod`)
Sometimes we need a function inside a class that doesn't use any data from the object (doesn't use `self`). For such cases, we use the `@staticmethod` decorator. It allows us to call the method without passing the instance automatically.

In [ ]:
class Calculator:
    def __init__(self, num):
        self.num = num

    @staticmethod
    def greet():
        print("Hello! Welcome to the Calculator.")

# We can call static methods without creating an object,
# or using an object without passing 'self'
Calculator.greet()

calc = Calculator(10)
calc.greet()

Hello! Welcome to the Calculator.
Hello! Welcome to the Calculator.


## Practice Questions

1. **Programmer Data Storage**
Create a class `Programmer` for storing information of programmers working at Microsoft. Include attributes like `name`, `product`, and `salary`.

In [ ]:
class Programmer:
    company = "Microsoft"
    def __init__(self, name, product, salary):
        self.name = name
        self.product = product
        self.salary = salary

p1 = Programmer("Sameer", "Skype", 1000000)
p2 = Programmer("Rohan", "Azure", 1200000)

print(f"Name: {p1.name}, Product: {p1.product}, Company: {p1.company}")
print(f"Name: {p2.name}, Product: {p2.product}, Company: {p2.company}")

Name: Sameer, Product: Skype, Company: Microsoft
Name: Rohan, Product: Azure, Company: Microsoft


2. **Calculator Class**
Write a class `Calculator` capable of finding square, cube, and square root of a number.

In [ ]:
class Calculator:
    def __init__(self, num):
        self.num = num

    def square(self):
        print(f"Square: {self.num ** 2}")

    def cube(self):
        print(f"Cube: {self.num ** 3}")

    def square_root(self):
        print(f"Square Root: {self.num ** 0.5}")

a = Calculator(9)
a.square()
a.cube()
a.square_root()

Square: 81
Cube: 729
Square Root: 3.0


3. **Attribute Precedence**
Create a class with a class attribute `a`. Create an object and set an instance attribute `a` directly. Does this change the class attribute?

In [ ]:
class Demo:
    a = "Class Level"

obj = Demo()
obj.a = "Instance Level"
print(f"Instance a: {obj.a}")
print(f"Class a: {Demo.a}")

Instance a: Instance Level
Class a: Class Level


4. **Train Reservation System**
Write a class `Train` with methods to book a ticket, get status (no of seats), and get fare information.

In [ ]:
class Train:
    def __init__(self, name, fare, seats):
        self.name = name
        self.fare = fare
        self.seats = seats

    def getStatus(self):
        print(f"Train: {self.name}, Seats available: {self.seats}")

    def getFareInfo(self):
        print(f"Fare: Rs. {self.fare}")

    def bookTicket(self):
        if(self.seats > 0):
            print(f"Booking successful! Seat number: {self.seats}")
            self.seats -= 1
        else:
            print("Sorry, the train is full!")

intercity = Train("Intercity Express", 90, 2)
intercity.getStatus()
intercity.bookTicket()
intercity.bookTicket()
intercity.bookTicket()

Train: Intercity Express, Seats available: 2
Booking successful! Seat number: 2
Booking successful! Seat number: 1
Sorry, the train is full!


5. **Self-Parameter Modification**
Can you change the name of the `self` parameter inside a class? Try renaming it and see if the code runs.

In [ ]:
class Sample:
    def __init__(slf, name):
        slf.name = name

obj = Sample("Test Name")
print(obj.name) # Renamed parameter to 'slf'. It works because 'self' is a convention.

Test Name


# Chapter 11 `INHERITANCE & ADVANCED OOP`

### 11.1 What is Inheritance?
Inheritance is a way of creating a new class from an existing one. The new class (Child/Derived class) inherits the attributes and methods of the existing class (Parent/Base class).

**Benefits:**
- Code Reusability.
- Logical Structure.

**Syntax:**
```python
class Parent:
    # Parent body

class Child(Parent):
    # Child body inherits Parent
```

In [ ]:
class Employee:
    company = "Google"
    def showDetails(self):
        print("This is an employee")

class Programmer(Employee):
    language = "Python"
    def getLanguage(self):
        print(f"The language is {self.language}")

p = Programmer()
print(p.company) # Inherited attribute
p.showDetails()  # Inherited method
p.getLanguage() # Own method

Google
This is an employee
The language is Python


### 11.2 Types of Inheritance
1. **Single Inheritance:** Derived class inherits from a single base class.
2. **Multiple Inheritance:** Derived class inherits from more than one base class.
3. **Multilevel Inheritance:** A derived class becomes the base class for another class.

In [ ]:
class Father:
    def trait1(self):
        print("Focus from Father")

class Mother:
    def trait2(self):
        print("Creativity from Mother")

class Child(Father, Mother):
    def show(self):
        print("Child inherits from both")

c = Child()
c.trait1()
c.trait2()
c.show()

Focus from Father
Creativity from Mother
Child inherits from both


### 11.3 The `super()` Method
The `super()` method is used to call methods of the parent class inside a child class. It is commonly used within `__init__` to ensure the parent class is properly initialized.

In [ ]:
class Person:
    def __init__(self):
        print("Initializing Person")

class Employee(Person):
    def __init__(self):
        super().__init__() # Calls Parent constructor
        print("Initializing Employee")

e = Employee()

Initializing Person
Initializing Employee


### 11.4 Class Methods (`@classmethod`)
A class method is bound to the class and not the object. It can modify a class state that applies across all instances. It takes `cls` as the first argument instead of `self`.

In [ ]:
class Employee:
    salary = 100

    @classmethod
    def changeSalary(cls, sal):
        cls.salary = sal

e = Employee()
print(f"Old Salary: {Employee.salary}")
e.changeSalary(200)
print(f"New Salary: {Employee.salary}")

Old Salary: 100
New Salary: 200


## Practice Questions

1. **2D and 3D Vectors**
Create a class `C2dVec` and use it to create another class `C3dVec` representing a 3-dimensional vector using inheritance.

In [ ]:
class C2dVec:
    def __init__(self, i, j):
        self.icap = i
        self.jcap = j

class C3dVec(C2dVec):
    def __init__(self, i, j, k):
        super().__init__(i, j)
        self.kcap = k

v2 = C2dVec(1, 3)
v3 = C3dVec(1, 9, 7)
print(f"2D: {v2.icap}i + {v2.jcap}j")
print(f"3D: {v3.icap}i + {v3.jcap}j + {v3.kcap}k")

2D: 1i + 3j
3D: 1i + 9j + 7k


2. **Electronic Device Inheritance**
Create a class `ElectronicDevice` and a class `Phone` that inherits from it. Then create a class `Smartphone` that inherits from `Phone`. Demonstrate multilevel inheritance by calling methods from all levels.

In [ ]:
class ElectronicDevice:
    def power_on(self):
        print("Device is turning on...")

# Level 2: Phone inherits from ElectronicDevice
class Phone(ElectronicDevice):
    def call(self):
        print("Making a phone call...")

# Level 3: Smartphone inherits from Phone
class Smartphone(Phone):
    def browse_internet(self):
        print("Browsing the web...")

# Create a Smartphone object
my_phone = Smartphone()

# Demonstrating multilevel inheritance:
my_phone.power_on()        # Inherited from ElectronicDevice
my_phone.call()            # Inherited from Phone
my_phone.browse_internet() # Defined in Smartphone

Device is turning on...
Making a phone call...
Browsing the web...


3. **Salary Increment Class Method**
Create a class `Employee` and add an instance variable `salary` and a class variable `increment`. Write a method `salaryAfterIncrement` which returns the total salary.

In [ ]:
class Employee:
    salary = 1000
    increment = 1.5

    @property
    def salaryAfterIncrement(self):
        return self.salary * self.increment

e = Employee()
print(f"Total Salary: {e.salaryAfterIncrement}")

Total Salary: 1500.0


4. **Complex Numbers**
Write a class `Complex` to represent complex numbers, along with overloaded operators `+` and `*` which adds and multiplies them.

In [ ]:
class Complex:
    def __init__(self, r, i):
        self.real = r
        self.imaginary = i

    def __add__(self, c):
        return Complex(self.real + c.real, self.imaginary + c.imaginary)

    def __str__(self):
        return f"{self.real} + {self.imaginary}i"

c1 = Complex(3, 2)
c2 = Complex(1, 7)
print(f"Sum: {c1 + c2}")

Sum: 4 + 9i


# Chapter 12 `ADVANCED PYTHON 1`

### 12.1 Modern Python Features
Python is constantly evolving. Here are some of the most useful features added in recent versions that make your code shorter and more readable.

#### 1. The Walrus Operator (`:=`)
Introduced in Python 3.8, it allows you to **assign** a value to a variable **inside** an expression (like an `if` statement or a `while` loop).

In [ ]:
# Example: Check length and store it at the same time
if (n := len("Python")) > 3:
    print(f"The word has {n} characters.")

The word has 6 characters.


#### 2. Type Hints
Type hints help other developers (and your code editor) understand what kind of data a function expects and returns. It doesn't stop the code from running with wrong types, but it makes it much easier to debug.

In [ ]:
def calculate_total(price: float, tax: float) -> float:
    return price + tax

print(calculate_total(100.0, 18.5))

118.5


#### 3. Match Case (Structural Pattern Matching)
Added in Python 3.10, this is like a more powerful version of `if-elif-else` for matching specific patterns or values.

In [ ]:
def get_day_type(day):
    match day:
        case "Saturday" | "Sunday":
            return "Weekend"
        case "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday":
            return "Weekday"
        case _:
            return "Invalid day"

print(get_day_type("Sunday"))

Weekend


### 12.2 Exception Handling
Exceptions are errors that happen during execution. Instead of letting your program crash, you can 'catch' these errors.

In [ ]:
try:
    num = int(input("Enter a number to divide 100: "))
    result = 100 / num
    print(f"Result: {result}")
except ZeroDivisionError:
    print("Error: You cannot divide by zero!")
except ValueError:
    print("Error: Please enter a valid integer.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")
finally:
    print("This block always runs, no matter what.")

Enter a number to divide 100: 3
Result: 33.333333333333336
This block always runs, no matter what.


### 12.3 List Comprehensions
This is a concise way to create lists from existing iterables. It's often faster and more 'Pythonic' than using a standard `for` loop.

In [ ]:
numbers = [1, 2, 3, 4, 5, 6]
# Create a list of squares for even numbers only
even_squares = [x**2 for x in numbers if x % 2 == 0]
print(even_squares) # Output: [4, 16, 36]

[4, 16, 36]


### 12.4 Enumerate & Global
- **Enumerate**: Gives you both the index and the value of items in a list during a loop.
- **Global**: Allows you to modify a variable defined outside the function.

In [ ]:
fruits = ["Apple", "Banana", "Cherry"]
for index, fruit in enumerate(fruits):
    print(f"{index}: {fruit}")

x = 10
def change_global():
    global x
    x = 20

change_global()
print(f"Value of x: {x}")

0: Apple
1: Banana
2: Cherry
Value of x: 20


## Practice Questions

1. **Walrus Operator Usage**

Write a program that keeps asking the user for input and prints its length, but stops when the user types 'quit'. Use the walrus operator inside the `while` loop condition.

In [ ]:

while (user_input := input("Enter something (type 'quit' to exit): ")) != "quit":
    print(f"Length of your input: {len(user_input)}")

Enter something (type 'quit' to exit): hey
Length of your input: 3
Enter something (type 'quit' to exit): how are you 
Length of your input: 12
Enter something (type 'quit' to exit): where are you
Length of your input: 13
Enter something (type 'quit' to exit): quit


2. **Type Hinted Function**

Create a function `get_average` that takes a list of integers and returns a float. Use proper type hinting.

In [ ]:
from typing import List

def get_average(numbers: List[int]) -> float:
    return sum(numbers) / len(numbers) if numbers else 0.0

print(get_average([10, 20, 30]))

20.0


3. **List Comprehension Filtering**

Given a list of names, create a new list containing only those names that have more than 5 characters using list comprehension.

`names = ["Alice", "Bob", "Charlie", "David", "Eve", "Franklin"]`

In [ ]:
names = ["Alice", "Bob", "Charlie", "David", "Eve", "Franklin"]
long_names = [name for name in names if len(name) > 5]
print(long_names)

['Charlie', 'Franklin']


4. **Exception Handling for Files**

Write a program that tries to open a file called `non_existent.txt`. Handle the `FileNotFoundError` gracefully by printing a friendly message instead of crashing.

In [ ]:
try:
    with open("non_existent.txt", "r") as f:
        content = f.read()
except FileNotFoundError:
    print("Sorry, the requested file was not found.")

Sorry, the requested file was not found.


5. **Match Case Status Codes**

Write a function `http_error` that takes an integer status code and returns a string description using `match case` (e.g., 200 -> "OK", 404 -> "Not Found", 500 -> "Server Error").

In [ ]:
def http_error(status):
    match status:
        case 200:
            return "OK"
        case 404:
            return "Not Found"
        case 500:
            return "Server Error"
        case _:
            return "Unknown Status"

print(http_error(404))

Not Found
